File: Sources/ThesisSkipApp/extensions/Color.swift
//
//  Color.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 03/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

import Foundation
import SwiftUI

//Code from: https://stackoverflow.com/a/56874327
extension Color {
    static func from(hex: String) -> Color {
        var cleanHex = hex
        if cleanHex.hasPrefix("#") {
            cleanHex = String(cleanHex.dropFirst())
        }

        guard let intValue = parseHex(cleanHex) else {
            return .black
        }

        var a: Int64 = 0
        var r: Int64 = 0
        var g: Int64 = 0
        var b: Int64 = 0
        switch cleanHex.count {
        case 3:  // RGB (12-bit)
            (a, r, g, b) = (
                Int64(255), (intValue >> 8) * 17, (intValue >> 4 & 0xF) * 17,
                (intValue & 0xF) * 17
            )
        case 6:  // RGB (24-bit)
            (a, r, g, b) = (
                Int64(255), intValue >> 16, intValue >> 8 & 0xFF, intValue & 0xFF
            )
        case 8:  // ARGB (32-bit)
            (a, r, g, b) = (
                intValue >> 24, intValue >> 16 & 0xFF, intValue >> 8 & 0xFF,
                intValue & 0xFF
            )
        default:
            (a, r, g, b) = (Int64(255), Int64(0), Int64(0), Int64(0))
        }

        return Color(
            Color.RGBColorSpace.sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue: Double(b) / 255,
            opacity: Double(a) / 255
        )
    }
    
    private static func parseHex(_ hex: String) -> Int64? {
            #if SKIP
            return hex.toLongOrNull(16)

            #else
            return Int64(hex, radix: 16)
            #endif
        }
}

File: .build/plugins/outputs/thesis-skip-app/ThesisSkipApp/destination/skipstone/ThesisSkipApp/src/main/kotlin/thesis/skip/app/Color.kt
//
//  Color.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 03/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

package thesis.skip.app

import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.Saver
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import skip.lib.*

import skip.foundation.*
import skip.ui.*
import skip.model.*

internal fun Color.Companion.from(hex: String): Color {
    var cleanHex = hex
    if (cleanHex.hasPrefix("#")) {
        cleanHex = String(cleanHex.dropFirst())
    }
    val intValue_0 = parseHex(cleanHex)
    if (intValue_0 == null) {
        return Color.black
    }

    var a: Long = 0
    var r: Long = 0
    var g: Long = 0
    var b: Long = 0
    when (cleanHex.count) {
        3 -> for (unusedi in 0..0) { val tmptuple = Tuple4(255L, (intValue_0 shr 8) * 17, (intValue_0 shr 4 and 0xF) * 17, (intValue_0 and 0xF) * 17); a = tmptuple.element0; r = tmptuple.element1; g = tmptuple.element2; b = tmptuple.element3 }
        6 -> for (unusedi in 0..0) { val tmptuple = Tuple4(255L, intValue_0 shr 16, intValue_0 shr 8 and 0xFF, intValue_0 and 0xFF); a = tmptuple.element0; r = tmptuple.element1; g = tmptuple.element2; b = tmptuple.element3 }
        8 -> for (unusedi in 0..0) { val tmptuple = Tuple4(intValue_0 shr 24, intValue_0 shr 16 and 0xFF, intValue_0 shr 8 and 0xFF, intValue_0 and 0xFF); a = tmptuple.element0; r = tmptuple.element1; g = tmptuple.element2; b = tmptuple.element3 }
        else -> for (unusedi in 0..0) { val tmptuple = Tuple4(255L, 0L, 0L, 0L); a = tmptuple.element0; r = tmptuple.element1; g = tmptuple.element2; b = tmptuple.element3 }
    }

    return Color(Color.RGBColorSpace.sRGB, red = Double(r) / 255, green = Double(g) / 255, blue = Double(b) / 255, opacity = Double(a) / 255)
}

private fun Color.Companion.parseHex(hex: String): Long? {
    return hex.toLongOrNull(16)

}
